enhancements/insights/conditional-data-gathering.md

```diff
@@ -0,0 +1,236 @@
+---
+title: conditional-data-gathering
+authors:
+  - "@Sergey1011010"
+reviewers:
+  - "@inecas"
+  - "@tremes"
+  - "@smarterclayton"
+approvers:
+  - "@smarterclayton"
+creation-date: 2021-07-15
+last-updated: 2021-07-15
+status: implementable
+---
+
+# Conditional Data Gathering
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [x] Design details are appropriately documented from clear requirements
+- [x] Test plan is defined
+- [ ] Operational readiness criteria is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+The conditional gatherer for Insights Operator collects data according to the defined gathering rules*.
+Each rule contains one or more conditions such as "alert A is firing"
+and one or more gatherers with parameters such as "collect X lines of logs from containers in namespace N".
+Current version has these rules defined in the code and the proposal is to load them from an external source
+to make collection of new conditional data faster. It's NOT about running brand new code,
+but just calling the existing gatherers with different validated parameters, so the operator can't
+collect anything we don't expect.
+
+\* note that rule here and later has nothing to do with rules used to analyze data in archives written
+by CCX presentation team
+
+## Motivation
+
+Collecting data on some condition such as an alert is quite common pattern
+during the root-cause analysis and the ability to define the rules
+what to collect on which condition externally will help to do that faster.
+
+### Goals
+
+- Implement validation of the rules for conditional gatherer
+- Create a service taking gathering rules from the git repo and providing them through the API
+- Fetch the rules definitions from the service's API and apply them to the conditional gatherer
+
+### Non-Goals
+
+Empty
+
+## Proposal
+
+Right now we have conditional gatherer which can collect data based on some conditions
```

Comment:
Both this design and #821 include a form of rule that looks for characteristics of a cluster before making a decision about what to do. The rules are for different purposes, but it would be worth spending some time considering whether we could share a data structure (nested within the API) and ideally evaluation code, so that we have just 1 implementation. Perhaps if not as a first version, we could have a goal of converging? And if not, let's document our reasoning.

/cc @wking @sdodson

Reply:
Well ClusterCondition could potentially be shared, it's very similar, but not sure if we would benefit from it. About evaluation code it depends if we would have any common conditions, so far we have AlertIsFiring condition and ClusterVersionMatches will be implemented and you have PromQLClusterCondition. Not sure if it worth the efforts since it may complicate adding new conditions for both parties.

Reply:
The task to unify APIs was created https://issues.redhat.com/browse/CCXDEV-5878